### PR TITLE
change panic to failure when parser-stack depleted

### LINKED ISSUE #66: json::decode() panics on malformed serialized JSON
`json::decode()` has the signature: `pub fn decode<T: Decodable>(s: &str) -> DecodeResult<T>`, yet panics when fed malformed JSON (when presumably it should return the `Err` variant of `DecoderResult)`.

To reproduce:

```rust
extern crate "rustc-serialize" as rustc_serialize; 

use rustc_serialize::json;

#[derive(Debug, RustcDecodable)]
pub enum ChatEvent {
    Variant(i32)
}

fn main() {
    let serialized = "{\"variant\": \"Variant\", \"fields\": []}";

    match json::decode(serialized) {
        Ok(event) => {
            let event: ChatEvent = event;
            println!("Ok {:?}", event);
        }, 
        Err(e) => println!("Err {:?}", e)
    }
}
```

When run:

```
$ RUST_BACKTRACE=1 cargo run--verbose            
       Fresh rustc-serialize v0.3.0
       Fresh codec_test v0.0.1 (file:///home/wilsoniya/Devel/rust/codec_test)
     Running `target/codec_test`
thread '<main>' panicked at 'called `Option::unwrap()` on a `None` value', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/option.rs:362
stack backtrace:
   1:         0xd7d38ed0 - sys::backtrace::write::h36f27d57607c02a1zfA
   2:         0xd7d3c260 - panicking::on_panic::h9138d9a94c738b62F9I
   3:         0xd7d32e00 - rt::unwind::begin_unwind_inner::ha129cf244cc26e118PI
   4:         0xd7d338d0 - rt::unwind::begin_unwind_fmt::h3eb3af7ad971d12cLOI
   5:         0xd7d3c180 - rust_begin_unwind
   6:         0xd7d67120 - panicking::panic_fmt::hb0af434144b60e81Y9r
   7:         0xd7d67030 - panicking::panic::hbfce9629384270cda8r
   8:         0xd7d2f0b0 - option::Option<T>::unwrap::h17985508233427626339
   9:         0xd7d2f040 - json::Decoder::pop::h3baa3de0ffceb8e2LFh
  10:         0xd7d2f4b0 - json::Decoder...Decoder::read_i32::haa735559d0d3751cP6h
  11:         0xd7d07ca0 - serialize::i32.Decodable::decode::h14025026717791657052
  12:         0xd7d07c50 - fn(&mut rustc.serialize..json..Decoder) .> core..result..Result<i32, rustc.serialize..json..DecoderError> $u7b$rustc.serialize..serialize..Decodable..decode$u7d$::fn_pointer_shim.2798::h54fb51cf77ad7b78
  13:         0xd7d07bf0 - ops::F.FnMut<A>::call_mut::h15546216301823000430
  14:         0xd7d07b90 - ops::F.FnOnce<A>::call_once::h6394649898615469647
  15:         0xd7d07b30 - json::Decoder...Decoder::read_enum_variant_arg::h16179315386903127666
  16:         0xd7d078f0 - ChatEvent...rustc_serialize..Decodable::decode::closure.2790
  17:         0xd7cfa990 - json::Decoder...Decoder::read_enum_variant::h14667130271225086222
  18:         0xd7cfa8e0 - ChatEvent...rustc_serialize..Decodable::decode::closure.2179
  19:         0xd7cfa880 - json::Decoder...Decoder::read_enum::h12831975579321903567
  20:         0xd7cfa7e0 - ChatEvent...rustc_serialize..Decodable::decode::h13508651075621685618
  21:         0xd7cf64d0 - json::decode::h16234319560854349711
  22:         0xd7cf60a0 - main::hdcecd00cb3507a16nca
  23:         0xd7d40c40 - rust_try_inner
  24:         0xd7d40c30 - rust_try
  25:         0xd7d3dd90 - rt::lang_start::h95373964c3e0731fl4I
  26:         0xd7cf6370 - main
  27:         0xd6f08f50 - __libc_start_main
  28:         0xd7cf5d50 - <unknown>
  29:                0x0 - <unknown>
Process didn't exit successfully: `target/codec_test` (status=101)
```
which appears to be due to `Decoder::pop()` calling `unwrap()` on a value pop'd from a `Vec`: http://doc.rust-lang.org/src/serialize/json.rs.html#2092 i.e.:

```rust
impl Decoder {
    fn pop(&mut self) -> Json {
        self.stack.pop().unwrap()
    }
}
```

### PR DESCRIPTION
**Pull request by oli-obk**
closes #66

### CODE REVIEW

**Comment by wilsoniya** [comment]
LGTM :+1: 

I wasn't invited to review, but I'm the reporter of #66 and this appears to fix the original issue.

File: src/json.rs [hunk 1]
```diff
@@ -3847,6 +3848,18 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_bad_json_stack_depleted() {
+        use json;
+        #[derive(Debug, RustcDecodable)]
+		enum ChatEvent {
```
Comment (wilsoniya) [inline src/json.rs]: Strange indentation; un-indent two levels on 3855-3860.

Comment (oli-obk) [inline src/json.rs]: sorry, i tried out RustDT and it uses tabs out of some reason. fixing...

**Comment by alexcrichton** [comment]
Thanks!

**Pull request merged**